### PR TITLE
ASoC: SOF: ipc4: fixup dailink based on copier format 

### PR DESCRIPTION
**Pull request by bardliao**
Some codecs don't support 32 bit format. We can use that way to specify the BE format. And it also makes sense that BE uses the same format as FE dai.

### CODE REVIEW

File: sound/soc/sof/ipc4-pcm.c
Comment (ranj063) [inline sound/soc/sof/ipc4-pcm.c]: @bardliao is this check sufficient? or should we check if really all formats are the same instead? like here: https://github.com/thesofproject/linux/blob/c4cf3be8612578cff99a567f9ee6451377e45aa6/sound/soc/sof/ipc4-topology.c#L1352

File: sound/soc/sof/ipc4-pcm.c
Comment (plbossart) [inline sound/soc/sof/ipc4-pcm.c:566]: I must admit not following the explanations *at all*.

This code checks for the formats on the dai copier, not the host. So why does the commit message refer to the FE DAI?

In addition, I am struggling to understand the relationship, if any, between the input and output formats of a dai copier, and what would happen if there is more than one format specified. How would that work for the Maxim amplifier which cannot support 24 bits?

**Review by plbossart — COMMENTED** [review]
not fully following your line of thought @bardliao

File: sound/soc/sof/ipc4-pcm.c
Comment (bardliao) [reply sound/soc/sof/ipc4-pcm.c:566]: > This code checks for the formats on the dai copier, not the host. So why does the commit message refer to the FE DAI?

@plbossart Sorry for confusing you. What I am trying to do is to set the BE format as the DAI format that closest to the BE.

Take sof-tgl-rt711-rt1316-rt714.tplg for example,
sof-tgl-rt711-rt1316-rt714

It is alh-copier.SDW0-Playback.0's output format for Jack out PCM and alh-copier.SDW0-Capture.0's input format for Jack In PCM.

Maybe I should not call `alh-copier.SDW0-Playback.0` and `alh-copier.SDW0-Capture.0` as FE DAI.


> In addition, I am struggling to understand the relationship, if any, between the input and output formats of a dai copier, and what would happen if there is more than one format specified. How would that work for the Maxim amplifier which cannot support 24 bits?

We won't know which format will be used if there are more than one formats available in a copier at this moment. BE format is decided before copier formats are selected. Therefore, we can only fixup to set the BE format if there is only one available format in the DAI copier.
Regarding Maxim amplifier, we will need to either add a new topology with valid_bit_depth = 24, or change valid_bit_depth to 24 in the existing topologies, and make sure the DAI copier supports only one output format.
BTW, that is the reason that I used SOF_TKN_COMP_FORMAT to specify the BE format just like IPC3 does.

File: sound/soc/sof/ipc4-pcm.c
Comment (bardliao) [reply sound/soc/sof/ipc4-pcm.c]: > @bardliao is this check sufficient? or should we check if really all formats are the same instead? like here:
> 
> https://github.com/thesofproject/linux/blob/c4cf3be8612578cff99a567f9ee6451377e45aa6/sound/soc/sof/ipc4-topology.c#L1352

@ranj063 I thought about this. The question is that do we want to specify the BE format whenever there is only one format available or we just want to specify the BE format when num_output_formats = 1?

**Review by plbossart — DISMISSED** [review]

File: sound/soc/sof/ipc4-pcm.c
Comment (ranj063) [reply sound/soc/sof/ipc4-pcm.c]: @bardliao what about the case where the formats are the same but rates/channels (like when you have an src in the pipeline) arent?

Comment (bardliao) [reply sound/soc/sof/ipc4-pcm.c]: > @bardliao what about the case where the formats are the same but rates/channels (like when you have an src in the pipeline) arent?

@ranj063 My intention was to use `num_output_formats = 1` to identify the needed of setting BE format. But, yes, it makes sense to set BE format whenever there is single format available.

**Review by ranj063 — APPROVED** [review]

**Review by plbossart — APPROVED** [review]

**Comment by plbossart** [comment]
@bardliao I can't recall if there are any SOF/topology dependencies needed before merging this. If we are good to go, please go ahead and merge - or submit a new version without the spurious newline. Your call.

**Comment by bardliao** [comment]
> @bardliao I can't recall if there are any SOF/topology dependencies needed before merging this. If we are good to go, please go ahead and merge - or submit a new version without the spurious newline. Your call.

I will merge this and remove the spurious newline in the follow up commit.

**Pull request merged**
